OMCompiler/Compiler/BackEnd/Tearing.mo

```diff
@@ -320,23 +330,28 @@ algorithm
       BackendDAE.JacobianType jacType;
       Boolean mixedSystem;
 
-    case ((BackendDAE.EQUATIONSYSTEM(eqns=eindex, vars=vindx, jac=BackendDAE.FULL_JACOBIAN(ojac), jacType=jacType, mixedSystem=mixedSystem)), _, _, _) equation
-      true = BackendDAEUtil.getLinearfromJacType(jacType);
-      maxSize = Flags.getConfigInt(Flags.MAX_SIZE_LINEAR_TEARING);
-      if intGt(listLength(vindx), maxSize) and not
-         // always apply tearing if maxSize > 0 and dense matrices are used
-         (intGt(maxSize, 0) and stringEqual(Config.simCodeTarget(), "Cpp") and
-          stringEqual(Flags.getConfigString(Flags.MATRIX_FORMAT), "dense"))
+    // Tearing of linear systems
+    case ((BackendDAE.EQUATIONSYSTEM(eqns=eindex, vars=vindx, jac=BackendDAE.FULL_JACOBIAN(ojac), jacType=jacType, mixedSystem=mixedSystem)), _, _, _)
+      guard BackendDAEUtil.getLinearfromJacType(jacType) algorithm
+
+      // Check if component size > maxSizeLinearTearing
+      maxSize := Flags.getConfigInt(Flags.MAX_SIZE_LINEAR_TEARING);
+      if listLength(vindx) > maxSize and
+        not (stringEqual(Config.simCodeTarget(), "Cpp") and stringEqual(Flags.getConfigString(Flags.MATRIX_FORMAT), "dense"))
```

Comment:
You dropped the `maxSize > 0` condition, was that intentional? I suppose this option is to allow disabling tearing even if you have Cpp runtime or dense solvers by setting `maxSize = 0`.

Reply:
I moved the test to https://github.com/OpenModelica/OpenModelica/pull/6976/files#diff-024aab96a9ad9fde002f39761bf2b6a17c682af5f997cbc1ca860efabbb3527dR261

We only need to do it once and not for every component, and if `maxSize<0` the user has used the debug flags wrong and omc should stop completely.

Reply:
> ..., and if `maxSize<0` the user has used the debug flags wrong and omc should stop completely.

I guess these kinds of tests should be done before the backend even starts, or the frontend for that matter? :sweat_smile:

Reply:
Probably, but I'm not sure where the correct place would be

Reply:
ConfigFlags have a validOptions field, but currently it's only implemented for string flags. I guess it should be quite straightforward to expand it to also handle min/max for numerical flags.

Reply:
It is good that you moved it outside, but besides that you don't allow negative numbers, you also changed behavior:

before:
`(-inf, 0]` -> no tearing
`(0, inf)` -> tearing

now:
`(-inf, 0)` -> error
`[0, inf)` -> tearing

my proposal:
`(-inf, 0)` -> error
`0` -> no tearing
`(0, inf)` -> tearing

The important part is, that you can disable tearing with setting `maxSize = 0` and i think that is intended.

Reply:
Ah scrap that, i forgot that you check for `intGt(listLength(vindx), maxSize)` which basically implies `maxSize > 0`.

All fine! :)

Reply:
My mind is going crazy here. I revoked my thoughts multiple times in my head now so here i go summarizing:

we have the condition:
`if listLength(vindx) > maxSize and
        not (stringEqual(Config.simCodeTarget(), "Cpp") and stringEqual(Flags.getConfigString(Flags.MATRIX_FORMAT), "dense"))`

which is true only if there are more variables than max size and either CPP or dense matrices are not used. Originally it also had to be the case that `maxSize = 0`can make this statement false. So by taking away that condition you make it impossible to switch off tearing if either CPP runtime or dense matrices are used by setting `maxSize = 0`. The intention here is that  `maxSize = 0` always turns of tearing, no matter the other settings.

Reply:
I should rewrite this completely. It is very hard to understand what should happen.

Reply:
Okay my last suggestion here would be this, then i am done:

```
  Boolean forceTearing = ((isCPP) or (isDense)) and not (maxSize == 0);

  if (#VARS > maxSize) and not forceTearing then
     <ERROR>
  else
     <TEARING>
  end if;
```

Reply:
I did it again and now it's clear what's happening (at least for me).
I also reduced a lot of duplicate code.